### DISCUSSION TITLE
Using not_allowed and status_handler plugins together?

### DISCUSSION THREAD

**Discussion by jgarber623** [jeremyevans/roda#267] [General]
Following up on #266 (thanks again for the fix there!), I noticed some confusing(ish) behavior when using the not_allowed ([documentation](http://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/NotAllowed.html)) and status_handler ([documentation](http://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/StatusHandler.html)) plugins.

## Sample Code

Here's an example app demonstrating the behavior (using Roda from source at 400671c):

```rb
require 'roda'

class App < Roda
  plugin :not_allowed
  plugin :status_handler
  
  route do |r|
    r.root do
      '👋🏻'
    end
  end
  
  status_handler(404) do
    '🤷🏻‍♂️'
  end
  
  status_handler(405) do
    '🙅🏻‍♂️'
  end
end

run App.freeze.app
```

Some sample HTTP requests and their responses (issued using [HTTPie](https://httpie.io)):

<details>
<summary><b>Click to expand sample HTTP request/responses…</b></summary>

```
$ http http://localhost:9292     
HTTP/1.1 200 OK
Content-Length: 8
Content-Type: text/html

👋🏻

$ http http://localhost:9292/foo 
HTTP/1.1 404 Not Found
Content-Length: 17
Content-Type: text/html

🤷🏻‍♂️

$ http post http://localhost:9292
HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 3
Content-Type: text/html

GET
```

</details>

## Observed Behavior

The last HTTP request (a `POST` to the root URL), correctly returns a 405 Method Not Allowed response status but _does not_ pass through the status_handler plugin's block for that status code (which attempts to return a body of '🙅🏻‍♂️'). As far as I could tell, plugin order does not make a difference in response behavior.

**This may be intended and/or designed behavior.** From the status_handler plugin documentation (linked above):

> The `status_handler` plugin adds a status_handler method which sets a block that is called whenever a response with the relevant response code with an empty body would be returned.

> Before a block is called, any existing headers on the response will be cleared. So if you want to be sure the headers are set even in your block, you need to reset them in the block.

…and from the not_allowed plugin documentation (also linked above):

> In all cases where it uses a 405 response, it also sets the Allow header in the response to contain the request methods supported.

The Allow header is required in a 405 response which, as written, would cause a small problem were the response shuffled through the status_handler plugin (which strips headers, as noted).

## Desired (maybe?) Behavior

There may be better or different ways to do this, but in my head, I'd be able to write a small app like:

```rb
require 'json'
require 'roda'
require 'tilt'

class App < Roda
  # Routing plugins
  plugin :not_allowed
  plugin :status_handler
  plugin :type_routing
  
  # Rendering plugins
  plugin :render
  
  route do |r|
    r.root do
      r.json { { message: 'Hello!' }.to_json }
      
      '👋🏻'
    end
  end
  
  status_handler(404) do |r|
    r.json { { error: 'Sorry!' }.to_json }
    
    '🤷🏻‍♂️'
  end
  
  status_handler(405) do |r|
    r.json { { error: 'No way!' }.to_json }
    
    '🙅🏻‍♂️'
  end
end

run App.freeze.app
```

…with the corresponding responses:

<details>
<summary><b>Click to expand sample HTTP request/responses…</b></summary>

```
$ http http://localhost:9292                             
HTTP/1.1 200 OK
Content-Length: 8
Content-Type: text/html

👋🏻

$ http http://localhost:9292 Accept:application/json     
HTTP/1.1 200 OK
Content-Length: 29
Content-Type: application/json
Vary: Accept

{
    "message": "Hello!"
}

$ http http://localhost:9292/foo                         
HTTP/1.1 404 Not Found
Content-Length: 17
Content-Type: text/html

🤷🏻‍♂️

$ http http://localhost:9292/foo  Accept:application/json
HTTP/1.1 404 Not Found
Content-Length: 18
Content-Type: application/json
Vary: Accept

{
    "error": "Sorry!"
}

$ http post http://localhost:9292                        
HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 3
Content-Type: text/html

🤷🏻‍♂️

$ http post http://localhost:9292 Accept:application/json
HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 3
Content-Type: application/json
Vary: Accept

{
    "error": "No way!"
}
```

</details>

Note that the first four requests and responses are achievable using the status_handler and type_routing plugins as-is. The last two responses are aspirational.

## In conclusion…

So… thanks for reading this far! I appreciate any thoughts you have on the above and how it may be possible.

**Comment by jeremyevans** [comment]
Thanks for posting this.  There are three separate issues here.  Two are bugs in the not_allowed plugin.  One bug is that a body would be unintentionally set for some responses.  Another bug is that the `r.root` support committed yesterday did not immediately stop processing.  The third issue is that the status_handler plugin did not offer a way to keep existing headers, which doesn't work well when using the not_allowed plugin.

I've pushed changes to address all three issues.  For the status_handler plugin, for 405 responses, you'll want to do `status_handler(405, keep_headers: true)` to make sure the `Allow` header is kept.

Can you please try the master branch to see if it operates the way you want?

**Reply by jgarber623** [reply]
@jeremyevans Thanks (again!) for the really quick reply. Here's what I tested:

## Application Code

```rb
# Gemfile

source 'https://rubygems.org'

ruby '3.1.2'

gem 'puma'
gem 'roda', github: 'jeremyevans/roda', ref: 'ef859de'
gem 'tilt'
```

```rb
# config.ru

require 'json'

require 'puma'
require 'roda'
require 'tilt'

class App < Roda
  # Routing plugins
  plugin :not_allowed
  plugin :status_handler
  plugin :type_routing
  
  # Rendering plugins
  plugin :render
  
  route do |r|
    r.root do
      r.json { { message: 'Okay!' }.to_json }
      
      '<h1>Okay!</h1>'
    end
  end
  
  status_handler(404) do |r|
    r.json { { message: 'Not found!' }.to_json }
    
    '<h1>Not found!</h1>'
  end
  
  status_handler(405, keep_headers: true) do |r|
    r.json { { message: 'Not allowed!' }.to_json }
    
    '<h1>Not allowed!</h1>'
  end
end

run App.freeze.app
```

## HTTPie Requests

```sh
# 200, app supports GET /
http http://localhost:9292
http http://localhost:9292 Accept:application/json

# 404, app does not support GET /foo
http http://localhost:9292/foo
http http://localhost:9292/foo Accept:application/json

# 405, app does not support POST /
http post http://localhost:9292
http post http://localhost:9292 Accept:application/json
```

The responses for those requests are _mostly_ correct.

<details>
<summary><b>Click to expand…</b></summary>

```
HTTP/1.1 200 OK
Content-Length: 14
Content-Type: text/html

<h1>Okay!</h1>


HTTP/1.1 200 OK
Content-Length: 19
Content-Type: application/json
Vary: Accept

{
    "message": "Okay!"
}


HTTP/1.1 404 Not Found
Content-Length: 19
Content-Type: text/html

<h1>Not found!</h1>


HTTP/1.1 404 Not Found
Content-Length: 24
Content-Type: application/json
Vary: Accept

{
    "message": "Not found!"
}


HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 21
Content-Type: text/html

<h1>Not allowed!</h1>


HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 26
Content-Type: text/html
Vary: Accept

{
    "message": "Not allowed!"
}
```

</details>

The only issue I see is the last request which POSTs to the root route and asks for JSON. The HTTP status code is correct (405), the body is correct (that JSON message), the `Vary: Accept` header is there, but the `Content-Type` header is `text/html` instead of `application/json`.

I _think_ that Content-Type header should be set automatically (maybe by the type_routing plugin?) but that could be conflicting with `keep_headers: true`…? That's a wild guess, though.

**Reply by jeremyevans** [reply]
The type_routing plugin doesn't override an existing Content-Type header. You probably want `response.headers.delete('Content-Type')` at the top of your 405 status handler.  Currently, only the `Content-Length` header is deleted if `:keep_headers` is used, because not doing can result in responses with invalid Content-Length (which violates the rack SPEC).

**Reply by jgarber623** [reply]
Adding `response.headers.delete('Content-Type')` as suggested did the trick. Thank you!

For posterity's sake, here's the updated `config.ru`:

```rb
require 'json'

require 'puma'
require 'roda'
require 'tilt'

class App < Roda
  # Routing plugins
  plugin :not_allowed
  plugin :status_handler
  plugin :type_routing
  
  # Rendering plugins
  plugin :render
  
  route do |r|
    r.root do
      r.json { { message: 'Okay!' }.to_json }
      
      '<h1>Okay!</h1>'
    end
  end
  
  status_handler(404) do |r|
    r.json { { message: 'Not found!' }.to_json }
    
    '<h1>Not found!</h1>'
  end
  
  status_handler(405, keep_headers: true) do |r|
    # Allow type_routing plugin to set the proper Content-Type header value
    response.headers.delete('Content-Type')
    
    r.json { { message: 'Not allowed!' }.to_json }
    
    '<h1>Not allowed!</h1>'
  end
end

run App.freeze.app
```

**Comment by jgarber623** [comment]
Regarding the status_handler plugin's new `keep_headers: true` option, here's an idea for an enhancement:

```rb
# As implemented, the option accepts a boolean
status_handler(405, keep_headers: true) {}

# Keep all headers (less 'Content-Length for conformance) except the
# headers provided in the array
status_handler(405, keep_headers: { except: ['Content-Type'] }) {}

# Keep only the headers provided and clear the rest (including 
# 'Content-Type' for conformance)
status_handler(405, keep_headers: { only: ['Accept'] }) {}
```

Admittedly, this complicates the underlying plugin implementation. There'd need to be a decision made about how to deconflict possible use cases (e.g. someone supplies _both_ `only` and `except` which could very quickly make no logical sense).

But… maybe worth considering?

**Reply by jeremyevans** [reply]
I suppose we could do: `keep_headers: ['Accept']`.  With an array, it would clear all but the given headers.  I'll implement that.

**Reply by jeremyevans** [reply]
OK, I implemented that in df6dff1b205d84b88dccb2fc802a7f7298025769.  I took out the previous behavior that only removed `Content-Length`, since I don't think it's behavior that most users would desire, due to `Content-Type` issues.

**Reply by jgarber623** [reply]
Excellent, I'll update and give this version a go. Thanks for considering the idea and implementing!

**Reply by jgarber623** [reply]
This version had the positive side effect of playing nicely with the default_headers plugin:

```rb
# config.ru
plugin :default_headers,
       'Content-Type' => 'text/html; charset=utf-8',
       'Referrer-Policy' => 'no-referrer-when-downgrade',
       'X-Frame-Options' => 'DENY',
       'X-XSS-Protection' => '0'
```

…and:

```sh
$ http post http://localhost:8080 Accept:application/json

HTTP/1.1 405 Method Not Allowed
Allow: GET
Content-Length: 49
Content-Type: application/json
Referrer-Policy: no-referrer-when-downgrade
Vary: Accept
X-Frame-Options: DENY
X-XSS-Protection: 0

{
    "message": "The requested method is not allowed"
}
```

🥳

**Comment by jgarber623** [comment]
@jeremyevans When might a new release be issued with these updates and the changes from #266?  No rush, of course! Looks like you've historically done one release about mid-month?

**Reply by jeremyevans** [reply]
The next release is planned for mid-next month.

**Reply by jgarber623** [reply]
Perfect, thanks!

**Comment by jgarber623** [comment]
Noting that this was released with v3.56.0 (#271). 🎉
